src/couch/src/couch_auto_purge_plugin.erl

```diff
@@ -0,0 +1,169 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_auto_purge_plugin).
+-behaviour(couch_scanner_plugin).
+
+-export([
+    start/2,
+    resume/2,
+    complete/1,
+    checkpoint/1,
+    db/2,
+    db_opened/2,
+    db_closing/2,
+    doc_fdi/3
+]).
+
+-include_lib("couch_scanner/include/couch_scanner_plugin.hrl").
+-include_lib("stdlib/include/assert.hrl").
+
+start(ScanId, #{}) ->
+    St = init_config(ScanId),
+    ?INFO("Starting.", [], St),
+    {ok, St}.
+
+resume(ScanId, #{}) ->
+    St = init_config(ScanId),
+    ?INFO("Resuming.", [], St),
+    {ok, St}.
+
+complete(St) ->
+    ?INFO("Completed", [], St),
+    {ok, #{}}.
+
+checkpoint(_St) ->
+    {ok, #{}}.
+
+db(St, DbName) ->
+    case ttl(St, DbName) of
+        TTL when is_integer(TTL) ->
+            {ok, St#{ttl => TTL}};
+        undefined ->
+            {skip, St}
+    end.
+
+db_opened(#{} = St, Db) ->
+    #{ttl := TTL} = St,
+    EndSeq = couch_time_seq:since(couch_db:get_time_seq(Db), couch_time_seq:timestamp() - TTL),
+    ChangeOpts =
+        if
+            EndSeq == now -> [];
+            true -> [{end_key, EndSeq}]
+        end,
+    ?INFO("scanning for deleted documents in ~s up to ~p", [couch_db:name(Db), EndSeq], meta(St)),
+    {0, ChangeOpts, St#{count => 0, end_seq => EndSeq}}.
+
+db_closing(#{} = St, Db) ->
+    #{count := Count} = St,
+    ?INFO("purged ~B deleted documents from ~s", [Count, couch_db:name(Db)], meta(St)),
+    {ok, St}.
+
+doc_fdi(#{} = St, #full_doc_info{deleted = true} = FDI, Db) ->
+    #{end_seq := EndSeq} = St,
+    ?assert(
+        FDI#full_doc_info.update_seq =< EndSeq, "FDI update_seq should not be greater than end seq"
+    ),
+    {ok, purge(St, FDI, Db)};
+doc_fdi(#{} = St, #full_doc_info{}, _Db) ->
+    {ok, St}.
+
+purge(#{} = St, #full_doc_info{} = FDI, Db) ->
```

Comment:
Related to a minor comment below we could get the `Id` in the function head `#full_doc_info{id = Id} = FDI` so then `fdi_to_revs` can just return the revs only this can be just a two liner with `purge(St, Id, fdi_to_revs(FD), MaxBatchSize, Db)`

But either way is fine, if this looks better to you, that's fine, too

Reply:
I feel that it's just a transformation of the full_doc_info record to another form, so just doing half (the more complicated half, admittedly) in a function and then stitching the two halves together at the calling site is just less tidy.